### PR TITLE
Individual questionnaire linked

### PR DESCRIPTION
**Pull request by AdeHarrison**
# Motivation and Context

# What has changed
Acceptance test added to check individual paper questionnaire request.
Improved "Questionnaire linked to unaddressed" test

# How to test?
[as per](https://github.com/ONSdigital/census-rm-case-processor/pull/53)

# Links
[Trello](https://trello.com/c/6mZA6Ynw/1040-rmc-142-part-2-handle-individual-pq-linked-event-8)

# Screenshots (if appropriate):

### CODE REVIEW

**Comment by AdeHarrison** [comment]
[RMC-142 (part 2) Handle Individual PQ Linked Event (8)](https://trello.com/c/6mZA6Ynw/1040-rmc-142-part-2-handle-individual-pq-linked-event-8)

**Review by gavinleeedwards — APPROVED** [review]
All tests pass against the same branch in case-processor.

File: acceptance_tests/features/steps/unaddressed.py [hunk 1]
```diff
@@ -65,8 +76,8 @@ def check_uac_message_is_received(context):
 
 @then("a Questionnaire Linked event is logged")
 def check_case_events(context):
-    case_id = context.linked_case['id']
-    time.sleep(2)  # Give case processor a chance to process the fulfilment request event
+    time.sleep(5)  # Give case processor a chance to process the Questionnaire Linked event
```
Comment (LukeLoze1) [inline acceptance_tests/features/steps/unaddressed.py]: I think we're trying to move away from using sleeps like this and have replaced them with retries
@retry(stop_max_attempt_number=10, wait_fixed=1000)

In this case I think the step would call another function with a retry on it

Comment (LukeLoze1) [reply acceptance_tests/features/steps/unaddressed.py]: suggestion for this area:  (code lifted from fulfilment.py) . function also called check_case_events, which here and there is a bit generic.  

@then("a Questionnaire Linked event is logged")
def check_questionaire_linked_logging(context):
    check_question_linked_event_is_logged(context)


@retry(stop_max_attempt_number=10, wait_fixed=1000)
def check_question_linked_event_is_logged(context):
    response = requests.get(f'{caseapi_url}{context.linked_case_id}', params={'caseEvents': True})
    assert 200 <= response.status_code <= 299, 'Get cases API call failed'
    cases = response.json()
    assert any(case_event['description'] == 'Questionnaire Linked' for case_event in cases['caseEvents'])

File: acceptance_tests/features/steps/unaddressed.py [hunk 2]
```diff
@@ -75,6 +86,14 @@ def check_case_events(context):
     assert False
 
 
+def _get_case_id_by_questionnaire_id(questionnaire_id):
+    time.sleep(5)  # Give case processor a chance to process the Questionnaire Linked event
```
Comment (LukeLoze1) [inline acceptance_tests/features/steps/unaddressed.py]: retry rather than sleep

**Review by LukeLoze1 — CHANGES_REQUESTED** [review]
some requested changes to remove some sleeps

File: acceptance_tests/features/steps/unaddressed.py (hunk 1 above)
Comment (AdeHarrison) [reply acceptance_tests/features/steps/unaddressed.py]: done

File: acceptance_tests/features/steps/unaddressed.py (hunk 2 above)
Comment (AdeHarrison) [reply acceptance_tests/features/steps/unaddressed.py]: done

File: acceptance_tests/features/unaddressed.feature [hunk 3]
```diff
@@ -7,9 +7,19 @@ Feature: Generating UAC/QID pairs for unaddressed letters & questionnaires
...
     And messages are emitted to RH and Action Scheduler with [01] questionnaire types
+    When an unaddressed message of questionnaire type 01 is sent
+    Then a UACUpdated message not linked to a case is emitted to RH and Action Scheduler
     When a Questionnaire Linked message is sent
     Then a Questionnaire Linked event is logged
 
+  Scenario: Individual Questionnaire linked to unaddressed
+    Given sample file "sample_for_questionnaire_linked.csv" is loaded
+    And messages are emitted to RH and Action Scheduler with [01] questionnaire types
+    When an unaddressed message of questionnaire type 21 is sent
```
Comment (liamtoozer) [inline acceptance_tests/features/unaddressed.feature:16]: Some of these When/Thens might need to be re-jigged to 'And' instead

File: acceptance_tests/features/unaddressed.feature [hunk 4]
```diff
@@ -7,9 +7,19 @@ Feature: Generating UAC/QID pairs for unaddressed letters & questionnaires
   Scenario: Questionnaire linked to unaddressed
     Given sample file "sample_for_questionnaire_linked.csv" is loaded
     And messages are emitted to RH and Action Scheduler with [01] questionnaire types
+    When an unaddressed message of questionnaire type 01 is sent
+    Then a UACUpdated message not linked to a case is emitted to RH and Action Scheduler
     When a Questionnaire Linked message is sent
     Then a Questionnaire Linked event is logged
 
+  Scenario: Individual Questionnaire linked to unaddressed
+    Given sample file "sample_for_questionnaire_linked.csv" is loaded
```
Comment (liamtoozer) [inline acceptance_tests/features/unaddressed.feature]: @AdamHawtin [mentioned in another PR](https://github.com/ONSdigital/census-rm-acceptance-tests/pull/121#discussion_r320592769) that these first 2 steps can be rewritten as 1 instead, using an existing step

**Review by LukeLoze1 — APPROVED** [review]
LGTM

File: acceptance_tests/features/unaddressed.feature (hunk 3 above)
Comment (liamtoozer) [reply acceptance_tests/features/unaddressed.feature:16]: Sorted in [this commit](https://github.com/ONSdigital/census-rm-acceptance-tests/pull/119/commits/a41158041a97cd30c910b713a559d2354af0ec91)

File: acceptance_tests/features/unaddressed.feature (hunk 4 above)
Comment (liamtoozer) [reply acceptance_tests/features/unaddressed.feature]: Sorted in [this commit](https://github.com/ONSdigital/census-rm-acceptance-tests/pull/119/commits/a41158041a97cd30c910b713a559d2354af0ec91)

**Review by liamtoozer — APPROVED** [review]
LGTM 👍

**Pull request merged**